{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# script that links the loops with the enhancers and how active they are. \n",
    "basically same as loop correlation but saves the results in a different format so I can annotate genes\n",
    "\n",
    "## to save time it will only do ATAC peaks near allelic imbalance and significant caQTL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 16 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import hicstraw \n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import glob\n",
    "import os\n",
    "import plotly.express as px\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors\n",
    "from pandarallel import pandarallel\n",
    "import cooler\n",
    "import cooltools\n",
    "import pybedtools as pbed\n",
    "pandarallel.initialize()\n",
    "from scipy import stats, special\n",
    "from statsmodels.stats import multitest\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import plotly.io as pio\n",
    "import seaborn as sns\n",
    "os.makedirs(\"/mnt/iusers01/jw01/mdefscs4/scratch/temp_pybedtools/\", exist_ok = True)\n",
    "pbed.helpers.set_tempdir(\"/mnt/iusers01/jw01/mdefscs4/scratch/temp_pybedtools/\")\n",
    "bed_genome_file = \"/mnt/iusers01/jw01/mdefscs4/hg38.genome\"\n",
    "\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "base_dir = \"/mnt/jw01-aruk-home01/projects/psa_functional_genomics/PsA_cleaned_analysis\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_annotation_df = pd.read_pickle(f\"{base_dir}/gencode_gtf.pickle\")\n",
    "gtf_transcripts = gtf_annotation_df[(gtf_annotation_df[\"feature\"] == \"transcript\") & (gtf_annotation_df[\"transcript_type\"] == \"protein_coding\")].dropna(axis=1, how='all')\n",
    "gtf_transcripts[\"gene_id\"] = gtf_transcripts[\"gene_id\"].str.split(\".\").str[0]\n",
    "gtf_transcripts[\"transcript_id\"] = gtf_transcripts[\"transcript_id\"].str.split(\".\").str[0]\n",
    "gtf_transcripts[\"TSS_start\"] = gtf_transcripts.apply(lambda x: int(x[\"start\"]) if x[\"strand\"] == \"+\" else int(x[\"end\"]) ,axis = 1)\n",
    "gtf_transcripts = gtf_transcripts[gtf_transcripts[\"seqname\"].str.startswith(\"chr\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA_normalized_counts = pd.read_csv(f\"{base_dir}/RNA_seq_analysis/RNA_normalized_counts.csv\")\n",
    "metadata_RNA = pd.read_csv(f\"{base_dir}/metadata/cleaned_RNA_metadata.csv\", index_col=0)\n",
    "column_name_dict = dict(zip(metadata_RNA['sample'], metadata_RNA['proper_name']))\n",
    "RNA_normalized_counts = RNA_normalized_counts.rename(columns=column_name_dict)\n",
    "RNA_normalized_counts_melted = pd.melt(RNA_normalized_counts, id_vars=[\"ensembl\",\"ENSG\",\"symbol\",\"genename\",\"entrez\"], \n",
    "        value_vars=RNA_normalized_counts.columns.difference([\"ensembl\",\"ENSG\",\"symbol\",\"genename\",\"entrez\"]),\n",
    "        var_name=\"proper_name\",value_name=\"expression\")\n",
    "RNA_normalized_counts_melted = RNA_normalized_counts_melted.merge(metadata_RNA[[\"patient\",\"cell_type\",\"condition\",\"proper_name\"]], on = \"proper_name\")\n",
    "\n",
    "# RNA_normalized_counts_melted = RNA_normalized_counts_melted[RNA_normalized_counts_melted[\"cell_type\"].isin([\"CD4\",\"CD8\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA_diff = pd.read_csv(f\"{base_dir}/RNA_seq_analysis/DEseq2_result/DE_CD8_vs_CD4_all.csv\", index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC_normalised_counts = pd.read_csv(f\"{base_dir}/ATAC_seq_analysis/ATAC_DESeq2_quantile_normalized_counts.csv\", index_col = 0)\n",
    "metadata_ATAC = pd.read_csv(f\"{base_dir}/metadata/cleaned_ATAC_metadata.csv\", index_col=0)\n",
    "column_name_dict = dict(zip(metadata_ATAC['id'], metadata_ATAC['proper_name']))\n",
    "ATAC_normalised_counts = ATAC_normalised_counts.rename(columns=column_name_dict)\n",
    "\n",
    "ATAC_normalised_counts_melted = pd.melt(ATAC_normalised_counts, id_vars=[\"CHR\",\"START\",\"END\"], \n",
    "        value_vars=ATAC_normalised_counts.columns.difference([\"CHR\",\"START\",\"END\"]),\n",
    "        var_name=\"proper_name\",value_name=\"peak_height\")\n",
    "ATAC_normalised_counts_melted = ATAC_normalised_counts_melted.merge(metadata_ATAC[[\"patient\",\"cell_type\",\"condition\",\"proper_name\"]], on=\"proper_name\", suffixes = [\"\", \"_y\"])\n",
    "\n",
    "# ATAC_normalised_counts_melted = ATAC_normalised_counts_melted[ATAC_normalised_counts_melted[\"cell_type\"].isin([\"CD4\",\"CD8\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC_diff = pd.read_csv(f\"{base_dir}/ATAC_seq_analysis/diffbind_result/DE_CD8_vs_CD4_all.csv\", index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "loops_analysed = pd.read_pickle(f\"{base_dir}/HiC_analysis/extracting_loop_counts/aggregated_counts/aggregated_normalized_loops_CD4_CD8.pk\")\n",
    "metadata_hic = pd.read_csv(f\"{base_dir}/metadata/cleaned_HiC_metadata.csv\", index_col = 0)\n",
    "column_name_dict = dict(zip(metadata_hic['folder_name'], metadata_hic['proper_name']))\n",
    "loops_analysed = loops_analysed.rename(columns=column_name_dict)\n",
    "loops_analysed[\"chrA\"] = \"chr\" + loops_analysed[\"chrA\"]\n",
    "loops_analysed[\"chrB\"] = \"chr\" + loops_analysed[\"chrB\"]\n",
    "loops_analysed = loops_analysed.reset_index()\n",
    "loops_counts_melted = pd.melt(loops_analysed, id_vars=['chrA', 'A_start', 'A_end', 'chrB', 'B_start', 'B_end', 'FDR', 'DETECTION_SCALE', 'distance_bin', 'index'], \n",
    "        value_vars=loops_analysed.columns.difference(['chrA', 'A_start', 'A_end', 'chrB', 'B_start', 'B_end', 'FDR', 'DETECTION_SCALE', 'distance_bin', 'index']),\n",
    "        var_name=\"proper_name\",value_name=\"interaction_strength\")\n",
    "\n",
    "loops_counts_melted = loops_counts_melted.merge(metadata_hic[[\"patient\",\"cell_type\",\"condition\",\"proper_name\"]], on = \"proper_name\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get the list of peaks that is a caQTL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC_permuted_CD8 = pd.read_csv(f\"{base_dir}/QTL_analysis/ATAC/output_final/ATAC_permuted_CD8_merged.txt\", sep = \" \")\n",
    "ATAC_permuted_CD4 = pd.read_csv(f\"{base_dir}/QTL_analysis/ATAC/output_final/ATAC_permuted_CD4_merged.txt\", sep = \" \")\n",
    "ATAC_permuted_CD8 = ATAC_permuted_CD8[ATAC_permuted_CD8[\"adj_beta_pval\"] < 0.1]\n",
    "ATAC_permuted_CD4 = ATAC_permuted_CD4[ATAC_permuted_CD4[\"adj_beta_pval\"] < 0.1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "ID_caQTL = set(ATAC_permuted_CD8[\"phe_id\"].to_list() + ATAC_permuted_CD4[\"phe_id\"].to_list())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get the list of peaks that are very near allelic imbalance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_SNPs_all = pd.read_pickle(f\"{base_dir}/ATAC_allelic_imbalance/combined_p_vals_files/all_SNPs_all.pkl\")\n",
    "all_SNPs_all = all_SNPs_all[(all_SNPs_all[\"corrected_p_val_greater\"] < 0.1) | (all_SNPs_all[\"corrected_p_val_less\"] < 0.1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps_anno_pbed = pbed.BedTool.from_dataframe(all_SNPs_all[[\"CHROM\",\"POS\",\"POS\",\"ID\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC_peaks = pbed.BedTool.from_dataframe(ATAC_normalised_counts.reset_index()[[\"CHR\",\"START\",\"END\", \"index\"]])\n",
    "\n",
    "res_anno_peaks = snps_anno_pbed.sort().closest(ATAC_peaks.sort(), d = True).to_dataframe(header = None, disable_auto_names = True)\n",
    "res_anno_peaks = res_anno_peaks[res_anno_peaks[8] < 500]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "ID_allelic_imb = set(res_anno_peaks[7].to_list())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "get all the peaks worth analysing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "ID_all = list(ID_caQTL.union(ID_allelic_imb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC_normalised_counts_filtered = ATAC_normalised_counts.loc[ID_all].copy()\n",
    "ATAC_normalised_counts_filtered = ATAC_normalised_counts_filtered.reset_index()\n",
    "ATAC_normalised_counts_melted = pd.melt(ATAC_normalised_counts_filtered, id_vars=[\"CHR\",\"START\",\"END\", \"index\"], \n",
    "        value_vars=ATAC_normalised_counts_filtered.columns.difference([\"CHR\",\"START\",\"END\", \"index\"]),\n",
    "        var_name=\"proper_name\",value_name=\"peak_height\")\n",
    "ATAC_normalised_counts_melted = ATAC_normalised_counts_melted.merge(metadata_ATAC[[\"patient\",\"cell_type\",\"condition\",\"proper_name\"]], on=\"proper_name\", suffixes = [\"\", \"_y\"])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## now we need to identify all the loops to test for each ATAC-peak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC_peaks = pbed.BedTool.from_dataframe(ATAC_normalised_counts_filtered[[\"CHR\",\"START\",\"END\", \"index\"]])\n",
    "loops = pbed.BedTool.from_dataframe(loops_analysed[['chrA', 'A_start', 'A_end', 'chrB', 'B_start', 'B_end', \"index\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "loops_with_peaks = loops.pairtobed(ATAC_peaks.slop(b=5000, g=bed_genome_file), type = \"either\").to_dataframe(header = None, disable_auto_names = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "loops_with_peaks.columns = ['chrA', 'A_start', 'A_end', 'chrB', 'B_start', 'B_end', \"index_loop\", \"CHR\",\"START\",\"END\", \"index_peak\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# correct the SLOP on the ATAC peaks\n",
    "loops_with_peaks[\"START\"] = loops_with_peaks[\"START\"] + 5000\n",
    "loops_with_peaks[\"END\"] = loops_with_peaks[\"END\"] - 5000"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## collect the p-values and regression coefficients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def retrieve_p_vals(peak_values, loop_values):\n",
    "    this_peak = peak_values.merge(loop_values[[\"proper_name\",\"interaction_strength\"]], on = \"proper_name\")\n",
    "    s = smf.ols(\"interaction_strength~peak_height\", data = this_peak) # interaction_strength~peak_height+cell_type+condition\n",
    "    r = s.fit()\n",
    "    return pd.Series({\"p_value\":r.pvalues.T[\"peak_height\"],\n",
    "                        \"beta\":r.params.T[\"peak_height\"],\n",
    "                        \"R_squared\":r.rsquared,\n",
    "                        \"R_squared_ajd\":r.rsquared_adj,\n",
    "                        \"interaction_mean\": loop_values[\"interaction_strength\"].mean(),\n",
    "                        \"peak_mean\":peak_values[\"peak_height\"].mean(),})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_regression(row):\n",
    "    index_peak = row[\"index_peak\"]\n",
    "    index_loop = row[\"index_loop\"]\n",
    "    peak_values = ATAC_normalised_counts_melted[ATAC_normalised_counts_melted[\"index\"] == index_peak]\n",
    "    loop_values = loops_counts_melted[loops_counts_melted[\"index\"] == index_loop]\n",
    "    res = retrieve_p_vals(peak_values, loop_values)\n",
    "    return pd.concat([row,res])\n",
    "loops_with_peaks = loops_with_peaks.parallel_apply(annotate_regression, axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "loops_with_peaks[\"FDR\"] = multitest.multipletests(loops_with_peaks[\"p_value\"], method = \"fdr_bh\",alpha = 0.1)[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "loops_with_peaks.to_pickle(f\"{base_dir}/integration_analysis/output_files/ATAC_corr_loops.pk\")\n",
    "loops_with_peaks.to_csv(f\"{base_dir}/integration_analysis/output_files/ATAC_corr_loops.csv.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "loops_with_peaks = pd.read_pickle(f\"{base_dir}/integration_analysis/output_files/ATAC_corr_loops.pk\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## now link genes to ATAC-peak and take significant ones "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "loops_with_peaks_bed = pbed.BedTool.from_dataframe(loops_with_peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_sites = gtf_transcripts[[\"seqname\", \"TSS_start\",\"gene_id\", \"gene_name\"]].copy()\n",
    "tss_sites[\"TSS_end\"] = tss_sites[\"TSS_start\"] + 1\n",
    "tss_sites = tss_sites.drop_duplicates([\"seqname\", \"TSS_start\",\"gene_id\", \"gene_name\"])\n",
    "tss_bed = pbed.BedTool.from_dataframe(tss_sites[[\"seqname\", \"TSS_start\", \"TSS_end\",\"gene_id\", \"gene_name\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "loops_with_peaks_genes = loops_with_peaks_bed.pairtobed(tss_bed.slop(b = 5000, g = bed_genome_file), type = \"either\").to_dataframe(header = None, disable_auto_names = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "loops_with_peaks_genes.columns = loops_with_peaks.columns.to_list() + [\"seqname\", \"TSS_start\", \"TSS_end\",\"gene_id\", \"gene_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove the genes that were basically overlapping the ATAC side, so keep only long range genes\n",
    "def drop_genes(row):\n",
    "    if (row[\"TSS_end\"] > row[\"START\"] - 5000) and (row[\"TSS_end\"] < row[\"END\"] + 5000):\n",
    "        return False\n",
    "    if (row[\"TSS_start\"] > row[\"START\"] -5000) and (row[\"TSS_start\"] < row[\"END\"] + 5000):\n",
    "        return False\n",
    "    return True\n",
    "rows_to_keep = loops_with_peaks_genes.apply(drop_genes, axis=1)\n",
    "loops_with_peaks_genes = loops_with_peaks_genes[rows_to_keep]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "significant_loops_with_peaks_genes = loops_with_peaks_genes[loops_with_peaks_genes[\"FDR\"] < 0.1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrA</th>\n",
       "      <th>A_start</th>\n",
       "      <th>A_end</th>\n",
       "      <th>chrB</th>\n",
       "      <th>B_start</th>\n",
       "      <th>B_end</th>\n",
       "      <th>index_loop</th>\n",
       "      <th>CHR</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>...</th>\n",
       "      <th>R_squared</th>\n",
       "      <th>R_squared_ajd</th>\n",
       "      <th>interaction_mean</th>\n",
       "      <th>peak_mean</th>\n",
       "      <th>FDR</th>\n",
       "      <th>seqname</th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>88504</th>\n",
       "      <td>chr2</td>\n",
       "      <td>110852500</td>\n",
       "      <td>110855000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111120000</td>\n",
       "      <td>111122500</td>\n",
       "      <td>14296</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858286</td>\n",
       "      <td>110858786</td>\n",
       "      <td>...</td>\n",
       "      <td>0.470858</td>\n",
       "      <td>0.465619</td>\n",
       "      <td>20.377721</td>\n",
       "      <td>772.688347</td>\n",
       "      <td>1.389225e-13</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111117670</td>\n",
       "      <td>111127671</td>\n",
       "      <td>ENSG00000153094</td>\n",
       "      <td>BCL2L11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88505</th>\n",
       "      <td>chr2</td>\n",
       "      <td>110852500</td>\n",
       "      <td>110855000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111120000</td>\n",
       "      <td>111122500</td>\n",
       "      <td>14296</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858286</td>\n",
       "      <td>110858786</td>\n",
       "      <td>...</td>\n",
       "      <td>0.470858</td>\n",
       "      <td>0.465619</td>\n",
       "      <td>20.377721</td>\n",
       "      <td>772.688347</td>\n",
       "      <td>1.389225e-13</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111118746</td>\n",
       "      <td>111128747</td>\n",
       "      <td>ENSG00000153094</td>\n",
       "      <td>BCL2L11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88506</th>\n",
       "      <td>chr2</td>\n",
       "      <td>110852500</td>\n",
       "      <td>110855000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111120000</td>\n",
       "      <td>111122500</td>\n",
       "      <td>14296</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858286</td>\n",
       "      <td>110858786</td>\n",
       "      <td>...</td>\n",
       "      <td>0.470858</td>\n",
       "      <td>0.465619</td>\n",
       "      <td>20.377721</td>\n",
       "      <td>772.688347</td>\n",
       "      <td>1.389225e-13</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111118752</td>\n",
       "      <td>111128753</td>\n",
       "      <td>ENSG00000153094</td>\n",
       "      <td>BCL2L11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88507</th>\n",
       "      <td>chr2</td>\n",
       "      <td>110852500</td>\n",
       "      <td>110855000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111120000</td>\n",
       "      <td>111122500</td>\n",
       "      <td>14296</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858286</td>\n",
       "      <td>110858786</td>\n",
       "      <td>...</td>\n",
       "      <td>0.470858</td>\n",
       "      <td>0.465619</td>\n",
       "      <td>20.377721</td>\n",
       "      <td>772.688347</td>\n",
       "      <td>1.389225e-13</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111114378</td>\n",
       "      <td>111124379</td>\n",
       "      <td>ENSG00000153094</td>\n",
       "      <td>BCL2L11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88508</th>\n",
       "      <td>chr2</td>\n",
       "      <td>110852500</td>\n",
       "      <td>110855000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111120000</td>\n",
       "      <td>111122500</td>\n",
       "      <td>14296</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858286</td>\n",
       "      <td>110858786</td>\n",
       "      <td>...</td>\n",
       "      <td>0.470858</td>\n",
       "      <td>0.465619</td>\n",
       "      <td>20.377721</td>\n",
       "      <td>772.688347</td>\n",
       "      <td>1.389225e-13</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111115914</td>\n",
       "      <td>111125915</td>\n",
       "      <td>ENSG00000153094</td>\n",
       "      <td>BCL2L11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88509</th>\n",
       "      <td>chr2</td>\n",
       "      <td>110852500</td>\n",
       "      <td>110855000</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111120000</td>\n",
       "      <td>111122500</td>\n",
       "      <td>14296</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858286</td>\n",
       "      <td>110858786</td>\n",
       "      <td>...</td>\n",
       "      <td>0.470858</td>\n",
       "      <td>0.465619</td>\n",
       "      <td>20.377721</td>\n",
       "      <td>772.688347</td>\n",
       "      <td>1.389225e-13</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111115929</td>\n",
       "      <td>111125930</td>\n",
       "      <td>ENSG00000153094</td>\n",
       "      <td>BCL2L11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrA    A_start      A_end  chrB    B_start      B_end  index_loop  \\\n",
       "88504  chr2  110852500  110855000  chr2  111120000  111122500       14296   \n",
       "88505  chr2  110852500  110855000  chr2  111120000  111122500       14296   \n",
       "88506  chr2  110852500  110855000  chr2  111120000  111122500       14296   \n",
       "88507  chr2  110852500  110855000  chr2  111120000  111122500       14296   \n",
       "88508  chr2  110852500  110855000  chr2  111120000  111122500       14296   \n",
       "88509  chr2  110852500  110855000  chr2  111120000  111122500       14296   \n",
       "\n",
       "        CHR      START        END  ...  R_squared  R_squared_ajd  \\\n",
       "88504  chr2  110858286  110858786  ...   0.470858       0.465619   \n",
       "88505  chr2  110858286  110858786  ...   0.470858       0.465619   \n",
       "88506  chr2  110858286  110858786  ...   0.470858       0.465619   \n",
       "88507  chr2  110858286  110858786  ...   0.470858       0.465619   \n",
       "88508  chr2  110858286  110858786  ...   0.470858       0.465619   \n",
       "88509  chr2  110858286  110858786  ...   0.470858       0.465619   \n",
       "\n",
       "       interaction_mean   peak_mean           FDR  seqname  TSS_start  \\\n",
       "88504         20.377721  772.688347  1.389225e-13     chr2  111117670   \n",
       "88505         20.377721  772.688347  1.389225e-13     chr2  111118746   \n",
       "88506         20.377721  772.688347  1.389225e-13     chr2  111118752   \n",
       "88507         20.377721  772.688347  1.389225e-13     chr2  111114378   \n",
       "88508         20.377721  772.688347  1.389225e-13     chr2  111115914   \n",
       "88509         20.377721  772.688347  1.389225e-13     chr2  111115929   \n",
       "\n",
       "         TSS_end          gene_id  gene_name  \n",
       "88504  111127671  ENSG00000153094    BCL2L11  \n",
       "88505  111128747  ENSG00000153094    BCL2L11  \n",
       "88506  111128753  ENSG00000153094    BCL2L11  \n",
       "88507  111124379  ENSG00000153094    BCL2L11  \n",
       "88508  111125915  ENSG00000153094    BCL2L11  \n",
       "88509  111125930  ENSG00000153094    BCL2L11  \n",
       "\n",
       "[6 rows x 23 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "significant_loops_with_peaks_genes[significant_loops_with_peaks_genes[\"gene_name\"] == \"BCL2L11\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "loops_with_peaks_genes.to_pickle(f\"{base_dir}/integration_analysis/output_files/ATAC_corr_loops_with_genes_ALL.pk\")\n",
    "loops_with_peaks_genes.to_csv(f\"{base_dir}/integration_analysis/output_files/ATAC_corr_loops_with_genes_ALL.csv\")\n",
    "\n",
    "significant_loops_with_peaks_genes.to_pickle(f\"{base_dir}/integration_analysis/output_files/ATAC_corr_loops_with_genes.pk\")\n",
    "significant_loops_with_peaks_genes.to_csv(f\"{base_dir}/integration_analysis/output_files/ATAC_corr_loops_with_genes.csv\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## finally just search the GWAS datasets\n",
    "this is just first pass, because the ideally we wanted the SNPs that have loopQTL or allelic imbalance, but this one doesn't filter for that, it just overlaps the peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_loops(file, name):\n",
    "    snp_bed = pbed.BedTool(file)\n",
    "    good_peaks = pbed.BedTool.from_dataframe(loops_with_peaks_genes[[\"CHR\",\"START\",\"END\", \"index_peak\"]].drop_duplicates())\n",
    "    snp_bed_with_peaks = snp_bed.intersect(good_peaks, wa = True, wb = True).to_dataframe(header = None, disable_auto_names=True)\n",
    "    snp_bed_with_peaks.columns = \"chr start end name score CHR START END index_peak\".split()\n",
    "    snp_bed_with_peaks[\"loci\"] = snp_bed_with_peaks[\"name\"].str.split(\"_\").str[-1]\n",
    "    snp_bed_with_peaks[\"snp\"] = snp_bed_with_peaks[\"name\"].str.split(\"_\").str[0]\n",
    "    SNPs_with_loop_with_gene = snp_bed_with_peaks[\"chr start end loci snp score index_peak\".split()].merge(loops_with_peaks_genes, on = \"index_peak\")\n",
    "    # filter allelic imbalance\n",
    "    SNPs_with_loop_with_gene = SNPs_with_loop_with_gene[SNPs_with_loop_with_gene[\"snp\"].isin(all_SNPs_all[\"ID\"])]\n",
    "    SNPs_with_loop_with_gene.to_csv(f\"{base_dir}/integration_analysis/output_files/ATAC_loop_genes_{name}.csv\")\n",
    "\n",
    "annotate_loops(f\"{base_dir}/metadata/GWAS_files/tsoi2017_LD_0.8_hg38.bed\", \"psoriasis_tsoi2017\")\n",
    "annotate_loops(f\"{base_dir}/metadata/GWAS_files/RAmetagwas_all_hg38.ld.bed\", \"RAmeta\")\n",
    "annotate_loops(f\"{base_dir}/metadata/GWAS_files/PsA_vs_controls_metagwas_significant.ld.hg38.bed\", \"PsA_meta\")\n",
    "annotate_loops(f\"{base_dir}/metadata/GWAS_files/suggestive_snps_hg38_ld.bed\", \"JIA_suggestive\")\n",
    "annotate_loops(f\"{base_dir}/metadata/GWAS_files/JIA_credible_snps_hg38.bed\", \"JIA_credible\")\n",
    "annotate_loops(f\"{base_dir}/metadata/GWAS_files/elena_hg38.ld.bed\", \"SSc_elena\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>loci</th>\n",
       "      <th>snp</th>\n",
       "      <th>score</th>\n",
       "      <th>index_peak</th>\n",
       "      <th>chrA</th>\n",
       "      <th>A_start</th>\n",
       "      <th>A_end</th>\n",
       "      <th>...</th>\n",
       "      <th>R_squared</th>\n",
       "      <th>R_squared_ajd</th>\n",
       "      <th>interaction_mean</th>\n",
       "      <th>peak_mean</th>\n",
       "      <th>FDR</th>\n",
       "      <th>seqname</th>\n",
       "      <th>TSS_start</th>\n",
       "      <th>TSS_end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67333211</td>\n",
       "      <td>67333212</td>\n",
       "      <td>rs4655698</td>\n",
       "      <td>rs11209051</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3723</td>\n",
       "      <td>chr1</td>\n",
       "      <td>67332500</td>\n",
       "      <td>67335000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.603381</td>\n",
       "      <td>0.599454</td>\n",
       "      <td>24.884601</td>\n",
       "      <td>297.983558</td>\n",
       "      <td>2.057858e-19</td>\n",
       "      <td>chr1</td>\n",
       "      <td>67425415</td>\n",
       "      <td>67435416</td>\n",
       "      <td>ENSG00000142864</td>\n",
       "      <td>SERBP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67333211</td>\n",
       "      <td>67333212</td>\n",
       "      <td>rs4655698</td>\n",
       "      <td>rs11209051</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3723</td>\n",
       "      <td>chr1</td>\n",
       "      <td>67332500</td>\n",
       "      <td>67335000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.603381</td>\n",
       "      <td>0.599454</td>\n",
       "      <td>24.884601</td>\n",
       "      <td>297.983558</td>\n",
       "      <td>2.057858e-19</td>\n",
       "      <td>chr1</td>\n",
       "      <td>67425386</td>\n",
       "      <td>67435387</td>\n",
       "      <td>ENSG00000142864</td>\n",
       "      <td>SERBP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67333211</td>\n",
       "      <td>67333212</td>\n",
       "      <td>rs4655698</td>\n",
       "      <td>rs11209051</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3723</td>\n",
       "      <td>chr1</td>\n",
       "      <td>67332500</td>\n",
       "      <td>67335000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.603381</td>\n",
       "      <td>0.599454</td>\n",
       "      <td>24.884601</td>\n",
       "      <td>297.983558</td>\n",
       "      <td>2.057858e-19</td>\n",
       "      <td>chr1</td>\n",
       "      <td>67425399</td>\n",
       "      <td>67435400</td>\n",
       "      <td>ENSG00000142864</td>\n",
       "      <td>SERBP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67333211</td>\n",
       "      <td>67333212</td>\n",
       "      <td>rs4655698</td>\n",
       "      <td>rs11209051</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3723</td>\n",
       "      <td>chr1</td>\n",
       "      <td>67332500</td>\n",
       "      <td>67335000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.603381</td>\n",
       "      <td>0.599454</td>\n",
       "      <td>24.884601</td>\n",
       "      <td>297.983558</td>\n",
       "      <td>2.057858e-19</td>\n",
       "      <td>chr1</td>\n",
       "      <td>67425372</td>\n",
       "      <td>67435373</td>\n",
       "      <td>ENSG00000142864</td>\n",
       "      <td>SERBP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>198614891</td>\n",
       "      <td>198614892</td>\n",
       "      <td>rs28398409</td>\n",
       "      <td>rs28398409</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>7724</td>\n",
       "      <td>chr1</td>\n",
       "      <td>198160000</td>\n",
       "      <td>198162500</td>\n",
       "      <td>...</td>\n",
       "      <td>0.060059</td>\n",
       "      <td>0.050753</td>\n",
       "      <td>12.508455</td>\n",
       "      <td>138.362658</td>\n",
       "      <td>7.029026e-02</td>\n",
       "      <td>chr1</td>\n",
       "      <td>198151963</td>\n",
       "      <td>198161964</td>\n",
       "      <td>ENSG00000151414</td>\n",
       "      <td>NEK7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr4</td>\n",
       "      <td>122578589</td>\n",
       "      <td>122578590</td>\n",
       "      <td>rs6814280</td>\n",
       "      <td>rs13140464</td>\n",
       "      <td>0.882487</td>\n",
       "      <td>76897</td>\n",
       "      <td>chr4</td>\n",
       "      <td>122577500</td>\n",
       "      <td>122580000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.151663</td>\n",
       "      <td>0.143263</td>\n",
       "      <td>10.051215</td>\n",
       "      <td>835.882068</td>\n",
       "      <td>6.222326e-04</td>\n",
       "      <td>chr4</td>\n",
       "      <td>122917442</td>\n",
       "      <td>122927443</td>\n",
       "      <td>ENSG00000170917</td>\n",
       "      <td>NUDT6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>chr4</td>\n",
       "      <td>122578589</td>\n",
       "      <td>122578590</td>\n",
       "      <td>rs6814280</td>\n",
       "      <td>rs13140464</td>\n",
       "      <td>0.882487</td>\n",
       "      <td>76897</td>\n",
       "      <td>chr4</td>\n",
       "      <td>122577500</td>\n",
       "      <td>122580000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.151663</td>\n",
       "      <td>0.143263</td>\n",
       "      <td>10.051215</td>\n",
       "      <td>835.882068</td>\n",
       "      <td>6.222326e-04</td>\n",
       "      <td>chr4</td>\n",
       "      <td>122918074</td>\n",
       "      <td>122928075</td>\n",
       "      <td>ENSG00000145375</td>\n",
       "      <td>SPATA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>chr5</td>\n",
       "      <td>56148855</td>\n",
       "      <td>56148856</td>\n",
       "      <td>rs7731626</td>\n",
       "      <td>rs7731626</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>79654</td>\n",
       "      <td>chr5</td>\n",
       "      <td>55985000</td>\n",
       "      <td>55990000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.765395</td>\n",
       "      <td>0.763073</td>\n",
       "      <td>47.579696</td>\n",
       "      <td>187.216611</td>\n",
       "      <td>5.159165e-30</td>\n",
       "      <td>chr5</td>\n",
       "      <td>55989993</td>\n",
       "      <td>55999994</td>\n",
       "      <td>ENSG00000134352</td>\n",
       "      <td>IL6ST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>chr5</td>\n",
       "      <td>56148855</td>\n",
       "      <td>56148856</td>\n",
       "      <td>rs7731626</td>\n",
       "      <td>rs7731626</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>79654</td>\n",
       "      <td>chr5</td>\n",
       "      <td>55985000</td>\n",
       "      <td>55990000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.765395</td>\n",
       "      <td>0.763073</td>\n",
       "      <td>47.579696</td>\n",
       "      <td>187.216611</td>\n",
       "      <td>5.159165e-30</td>\n",
       "      <td>chr5</td>\n",
       "      <td>55989964</td>\n",
       "      <td>55999965</td>\n",
       "      <td>ENSG00000134352</td>\n",
       "      <td>IL6ST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>chr5</td>\n",
       "      <td>56148855</td>\n",
       "      <td>56148856</td>\n",
       "      <td>rs7731626</td>\n",
       "      <td>rs7731626</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>79654</td>\n",
       "      <td>chr5</td>\n",
       "      <td>55985000</td>\n",
       "      <td>55990000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.765395</td>\n",
       "      <td>0.763073</td>\n",
       "      <td>47.579696</td>\n",
       "      <td>187.216611</td>\n",
       "      <td>5.159165e-30</td>\n",
       "      <td>chr5</td>\n",
       "      <td>55977800</td>\n",
       "      <td>55987801</td>\n",
       "      <td>ENSG00000134352</td>\n",
       "      <td>IL6ST</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr      start        end        loci         snp     score  index_peak  \\\n",
       "0   chr1   67333211   67333212   rs4655698  rs11209051  1.000000        3723   \n",
       "1   chr1   67333211   67333212   rs4655698  rs11209051  1.000000        3723   \n",
       "2   chr1   67333211   67333212   rs4655698  rs11209051  1.000000        3723   \n",
       "3   chr1   67333211   67333212   rs4655698  rs11209051  1.000000        3723   \n",
       "4   chr1  198614891  198614892  rs28398409  rs28398409  1.000000        7724   \n",
       "..   ...        ...        ...         ...         ...       ...         ...   \n",
       "80  chr4  122578589  122578590   rs6814280  rs13140464  0.882487       76897   \n",
       "81  chr4  122578589  122578590   rs6814280  rs13140464  0.882487       76897   \n",
       "82  chr5   56148855   56148856   rs7731626   rs7731626  1.000000       79654   \n",
       "83  chr5   56148855   56148856   rs7731626   rs7731626  1.000000       79654   \n",
       "84  chr5   56148855   56148856   rs7731626   rs7731626  1.000000       79654   \n",
       "\n",
       "    chrA    A_start      A_end  ... R_squared  R_squared_ajd  \\\n",
       "0   chr1   67332500   67335000  ...  0.603381       0.599454   \n",
       "1   chr1   67332500   67335000  ...  0.603381       0.599454   \n",
       "2   chr1   67332500   67335000  ...  0.603381       0.599454   \n",
       "3   chr1   67332500   67335000  ...  0.603381       0.599454   \n",
       "4   chr1  198160000  198162500  ...  0.060059       0.050753   \n",
       "..   ...        ...        ...  ...       ...            ...   \n",
       "80  chr4  122577500  122580000  ...  0.151663       0.143263   \n",
       "81  chr4  122577500  122580000  ...  0.151663       0.143263   \n",
       "82  chr5   55985000   55990000  ...  0.765395       0.763073   \n",
       "83  chr5   55985000   55990000  ...  0.765395       0.763073   \n",
       "84  chr5   55985000   55990000  ...  0.765395       0.763073   \n",
       "\n",
       "    interaction_mean   peak_mean           FDR  seqname  TSS_start    TSS_end  \\\n",
       "0          24.884601  297.983558  2.057858e-19     chr1   67425415   67435416   \n",
       "1          24.884601  297.983558  2.057858e-19     chr1   67425386   67435387   \n",
       "2          24.884601  297.983558  2.057858e-19     chr1   67425399   67435400   \n",
       "3          24.884601  297.983558  2.057858e-19     chr1   67425372   67435373   \n",
       "4          12.508455  138.362658  7.029026e-02     chr1  198151963  198161964   \n",
       "..               ...         ...           ...      ...        ...        ...   \n",
       "80         10.051215  835.882068  6.222326e-04     chr4  122917442  122927443   \n",
       "81         10.051215  835.882068  6.222326e-04     chr4  122918074  122928075   \n",
       "82         47.579696  187.216611  5.159165e-30     chr5   55989993   55999994   \n",
       "83         47.579696  187.216611  5.159165e-30     chr5   55989964   55999965   \n",
       "84         47.579696  187.216611  5.159165e-30     chr5   55977800   55987801   \n",
       "\n",
       "            gene_id  gene_name  \n",
       "0   ENSG00000142864     SERBP1  \n",
       "1   ENSG00000142864     SERBP1  \n",
       "2   ENSG00000142864     SERBP1  \n",
       "3   ENSG00000142864     SERBP1  \n",
       "4   ENSG00000151414       NEK7  \n",
       "..              ...        ...  \n",
       "80  ENSG00000170917      NUDT6  \n",
       "81  ENSG00000145375     SPATA5  \n",
       "82  ENSG00000134352      IL6ST  \n",
       "83  ENSG00000134352      IL6ST  \n",
       "84  ENSG00000134352      IL6ST  \n",
       "\n",
       "[61 rows x 29 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
