{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# this script allows to print all the information for a specific SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 28 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import hicstraw \n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import glob\n",
    "import os\n",
    "import plotly.express as px\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors\n",
    "from pandarallel import pandarallel\n",
    "import cooler\n",
    "import cooltools\n",
    "import pybedtools as pbed\n",
    "pandarallel.initialize()\n",
    "from scipy import stats, special\n",
    "from statsmodels.stats import multitest\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import plotly.io as pio\n",
    "import seaborn as sns\n",
    "\n",
    "from functools import reduce\n",
    "\n",
    "os.makedirs(\"/mnt/iusers01/jw01/mdefscs4/scratch/temp_pybedtools/\", exist_ok = True)\n",
    "pbed.helpers.set_tempdir(\"/mnt/iusers01/jw01/mdefscs4/scratch/temp_pybedtools/\")\n",
    "bed_genome_file = \"/mnt/iusers01/jw01/mdefscs4/hg38.genome\"\n",
    "\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "base_dir = \"/mnt/jw01-aruk-home01/projects/psa_functional_genomics/PsA_cleaned_analysis\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_annotation_df = pd.read_pickle(f\"{base_dir}/gencode_gtf.pickle\")\n",
    "gtf_transcripts = gtf_annotation_df[(gtf_annotation_df[\"feature\"] == \"transcript\") & (gtf_annotation_df[\"transcript_type\"] == \"protein_coding\")].dropna(axis=1, how='all')\n",
    "gtf_transcripts[\"gene_id\"] = gtf_transcripts[\"gene_id\"].str.split(\".\").str[0]\n",
    "gtf_transcripts[\"transcript_id\"] = gtf_transcripts[\"transcript_id\"].str.split(\".\").str[0]\n",
    "gtf_transcripts[\"TSS_start\"] = gtf_transcripts.apply(lambda x: int(x[\"start\"]) if x[\"strand\"] == \"+\" else int(x[\"end\"]) ,axis = 1)\n",
    "gene_mapper = dict(zip(gtf_transcripts['gene_id'], gtf_transcripts['gene_name']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA_nominal_CD4 = pd.read_csv(f\"{base_dir}/QTL_analysis/RNA/output_final/RNA_nominal_CD4_merged.txt\", sep = \" \")\n",
    "RNA_nominal_CD8 = pd.read_csv(f\"{base_dir}/QTL_analysis/RNA/output_final/RNA_nominal_CD8_merged.txt\", sep = \" \")\n",
    "ATAC_nominal_CD8 = pd.read_csv(f\"{base_dir}/QTL_analysis/ATAC/output_final/ATAC_nominal_CD8_merged.txt\", sep = \" \")\n",
    "ATAC_nominal_CD4 = pd.read_csv(f\"{base_dir}/QTL_analysis/ATAC/output_final/ATAC_nominal_CD4_merged.txt\", sep = \" \")\n",
    "ins_nominal_CD8 = pd.read_csv(f\"{base_dir}/QTL_analysis/HiC/output_final/ins_nominal_CD8_merged.txt\", sep = \" \")\n",
    "ins_nominal_CD4 = pd.read_csv(f\"{base_dir}/QTL_analysis/HiC/output_final/ins_nominal_CD4_merged.txt\", sep = \" \")\n",
    "loop_nominal_CD8 = pd.read_csv(f\"{base_dir}/QTL_analysis/HiC/output_final/loop_nominal_CD8_merged.txt\", sep = \" \")\n",
    "loop_nominal_CD4 = pd.read_csv(f\"{base_dir}/QTL_analysis/HiC/output_final/loop_nominal_CD4_merged.txt\", sep = \" \")\n",
    "\n",
    "output_dataframe_CD8 = pd.read_csv(f\"{base_dir}/HiC_allelic_imbalance/output_dataframe_CD8.csv\", index_col=0)\n",
    "output_dataframe_CD4 = pd.read_csv(f\"{base_dir}/HiC_allelic_imbalance/output_dataframe_CD4.csv\", index_col=0)\n",
    "\n",
    "o = f\"{base_dir}/ATAC_allelic_imbalance/combined_p_vals_files/\"\n",
    "all_SNPs_all = pd.read_pickle(o + \"all_SNPs_all.pkl\").drop(columns = [\"eQTLgen_gene\", \"snp\",\"hsc_genes\",\"tcell_genes\",\"all_genes\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA_nominal_CD4[\"gene_name\"] = RNA_nominal_CD4['phe_id'].map(gene_mapper)\n",
    "RNA_nominal_CD8[\"gene_name\"] = RNA_nominal_CD8['phe_id'].map(gene_mapper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_all_tables(vars):\n",
    "    if not type(vars) == list:\n",
    "        vars = [vars]\n",
    "    print(\"eQTL CD4\")\n",
    "    display(RNA_nominal_CD4[RNA_nominal_CD4[\"var_id\"].isin(vars)])\n",
    "\n",
    "    print(\"eQTL CD8\")\n",
    "    display(RNA_nominal_CD8[RNA_nominal_CD8[\"var_id\"].isin(vars)])\n",
    "\n",
    "    print(\"caQTL CD4\")\n",
    "    display(ATAC_nominal_CD4[ATAC_nominal_CD4[\"var_id\"].isin(vars)])\n",
    "\n",
    "    print(\"caQTL CD8\")\n",
    "    display(ATAC_nominal_CD8[ATAC_nominal_CD8[\"var_id\"].isin(vars)])\n",
    "\n",
    "    print(\"loopQTL CD4\")\n",
    "    display(loop_nominal_CD4[loop_nominal_CD4[\"var_id\"].isin(vars)])\n",
    "\n",
    "    print(\"loopQTL CD8\")\n",
    "    display(loop_nominal_CD8[loop_nominal_CD8[\"var_id\"].isin(vars)])\n",
    "\n",
    "    print(\"insQTL CD4\")\n",
    "    display(ins_nominal_CD4[ins_nominal_CD4[\"var_id\"].isin(vars)])\n",
    "\n",
    "    print(\"insQTL CD8\")\n",
    "    display(ins_nominal_CD8[ins_nominal_CD8[\"var_id\"].isin(vars)])\n",
    "\n",
    "\n",
    "    print(\"allelic imbalance\")\n",
    "    display(all_SNPs_all[all_SNPs_all[\"ID\"].isin(vars)])\n",
    "\n",
    "    print(\"loops with allelic imbalance CD4\")\n",
    "    display(output_dataframe_CD4[output_dataframe_CD4[\"rsID\"].isin(vars)])\n",
    "\n",
    "    print(\"loops with allelic imbalance CD8\")\n",
    "    display(output_dataframe_CD8[output_dataframe_CD8[\"rsID\"].isin(vars)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eQTL CD4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phe_id</th>\n",
       "      <th>phe_chr</th>\n",
       "      <th>phe_from</th>\n",
       "      <th>phe_to</th>\n",
       "      <th>phe_strd</th>\n",
       "      <th>n_var_in_cis</th>\n",
       "      <th>dist_phe_var</th>\n",
       "      <th>var_id</th>\n",
       "      <th>var_chr</th>\n",
       "      <th>var_from</th>\n",
       "      <th>var_to</th>\n",
       "      <th>nom_pval</th>\n",
       "      <th>r_squared</th>\n",
       "      <th>slope</th>\n",
       "      <th>best_hit</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [phe_id, phe_chr, phe_from, phe_to, phe_strd, n_var_in_cis, dist_phe_var, var_id, var_chr, var_from, var_to, nom_pval, r_squared, slope, best_hit, gene_name]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eQTL CD8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phe_id</th>\n",
       "      <th>phe_chr</th>\n",
       "      <th>phe_from</th>\n",
       "      <th>phe_to</th>\n",
       "      <th>phe_strd</th>\n",
       "      <th>n_var_in_cis</th>\n",
       "      <th>dist_phe_var</th>\n",
       "      <th>var_id</th>\n",
       "      <th>var_chr</th>\n",
       "      <th>var_from</th>\n",
       "      <th>var_to</th>\n",
       "      <th>nom_pval</th>\n",
       "      <th>r_squared</th>\n",
       "      <th>slope</th>\n",
       "      <th>best_hit</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [phe_id, phe_chr, phe_from, phe_to, phe_strd, n_var_in_cis, dist_phe_var, var_id, var_chr, var_from, var_to, nom_pval, r_squared, slope, best_hit, gene_name]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "caQTL CD4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phe_id</th>\n",
       "      <th>phe_chr</th>\n",
       "      <th>phe_from</th>\n",
       "      <th>phe_to</th>\n",
       "      <th>phe_strd</th>\n",
       "      <th>n_var_in_cis</th>\n",
       "      <th>dist_phe_var</th>\n",
       "      <th>var_id</th>\n",
       "      <th>var_chr</th>\n",
       "      <th>var_from</th>\n",
       "      <th>var_to</th>\n",
       "      <th>nom_pval</th>\n",
       "      <th>r_squared</th>\n",
       "      <th>slope</th>\n",
       "      <th>best_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>659007</th>\n",
       "      <td>55016</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858287</td>\n",
       "      <td>110858786</td>\n",
       "      <td>+</td>\n",
       "      <td>2469</td>\n",
       "      <td>0</td>\n",
       "      <td>rs13401811</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858527</td>\n",
       "      <td>110858527</td>\n",
       "      <td>0.000451</td>\n",
       "      <td>0.232495</td>\n",
       "      <td>-0.813462</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        phe_id phe_chr   phe_from     phe_to phe_strd  n_var_in_cis  \\\n",
       "659007   55016    chr2  110858287  110858786        +          2469   \n",
       "\n",
       "        dist_phe_var      var_id var_chr   var_from     var_to  nom_pval  \\\n",
       "659007             0  rs13401811    chr2  110858527  110858527  0.000451   \n",
       "\n",
       "        r_squared     slope  best_hit  \n",
       "659007   0.232495 -0.813462         0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "caQTL CD8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phe_id</th>\n",
       "      <th>phe_chr</th>\n",
       "      <th>phe_from</th>\n",
       "      <th>phe_to</th>\n",
       "      <th>phe_strd</th>\n",
       "      <th>n_var_in_cis</th>\n",
       "      <th>dist_phe_var</th>\n",
       "      <th>var_id</th>\n",
       "      <th>var_chr</th>\n",
       "      <th>var_from</th>\n",
       "      <th>var_to</th>\n",
       "      <th>nom_pval</th>\n",
       "      <th>r_squared</th>\n",
       "      <th>slope</th>\n",
       "      <th>best_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>694837</th>\n",
       "      <td>55016</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858287</td>\n",
       "      <td>110858786</td>\n",
       "      <td>+</td>\n",
       "      <td>2469</td>\n",
       "      <td>0</td>\n",
       "      <td>rs13401811</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858527</td>\n",
       "      <td>110858527</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.310338</td>\n",
       "      <td>-0.88706</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        phe_id phe_chr   phe_from     phe_to phe_strd  n_var_in_cis  \\\n",
       "694837   55016    chr2  110858287  110858786        +          2469   \n",
       "\n",
       "        dist_phe_var      var_id var_chr   var_from     var_to  nom_pval  \\\n",
       "694837             0  rs13401811    chr2  110858527  110858527  0.000001   \n",
       "\n",
       "        r_squared    slope  best_hit  \n",
       "694837   0.310338 -0.88706         0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loopQTL CD4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phe_id</th>\n",
       "      <th>phe_chr</th>\n",
       "      <th>phe_from</th>\n",
       "      <th>phe_to</th>\n",
       "      <th>phe_strd</th>\n",
       "      <th>n_var_in_cis</th>\n",
       "      <th>dist_phe_var</th>\n",
       "      <th>var_id</th>\n",
       "      <th>var_chr</th>\n",
       "      <th>var_from</th>\n",
       "      <th>var_to</th>\n",
       "      <th>nom_pval</th>\n",
       "      <th>r_squared</th>\n",
       "      <th>slope</th>\n",
       "      <th>best_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>593331</th>\n",
       "      <td>14298</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110855001</td>\n",
       "      <td>111465000</td>\n",
       "      <td>+</td>\n",
       "      <td>3555</td>\n",
       "      <td>0</td>\n",
       "      <td>rs13401811</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858527</td>\n",
       "      <td>110858527</td>\n",
       "      <td>0.002480</td>\n",
       "      <td>0.197913</td>\n",
       "      <td>-0.717554</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593575</th>\n",
       "      <td>14307</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111012501</td>\n",
       "      <td>111730000</td>\n",
       "      <td>+</td>\n",
       "      <td>4123</td>\n",
       "      <td>-153974</td>\n",
       "      <td>rs13401811</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858527</td>\n",
       "      <td>110858527</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.160169</td>\n",
       "      <td>-0.604286</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596047</th>\n",
       "      <td>14363</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111747501</td>\n",
       "      <td>111852500</td>\n",
       "      <td>+</td>\n",
       "      <td>4052</td>\n",
       "      <td>-888974</td>\n",
       "      <td>rs13401811</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858527</td>\n",
       "      <td>110858527</td>\n",
       "      <td>0.005610</td>\n",
       "      <td>0.168733</td>\n",
       "      <td>-0.662550</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        phe_id phe_chr   phe_from     phe_to phe_strd  n_var_in_cis  \\\n",
       "593331   14298    chr2  110855001  111465000        +          3555   \n",
       "593575   14307    chr2  111012501  111730000        +          4123   \n",
       "596047   14363    chr2  111747501  111852500        +          4052   \n",
       "\n",
       "        dist_phe_var      var_id var_chr   var_from     var_to  nom_pval  \\\n",
       "593331             0  rs13401811    chr2  110858527  110858527  0.002480   \n",
       "593575       -153974  rs13401811    chr2  110858527  110858527  0.007107   \n",
       "596047       -888974  rs13401811    chr2  110858527  110858527  0.005610   \n",
       "\n",
       "        r_squared     slope  best_hit  \n",
       "593331   0.197913 -0.717554         1  \n",
       "593575   0.160169 -0.604286         0  \n",
       "596047   0.168733 -0.662550         0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loopQTL CD8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phe_id</th>\n",
       "      <th>phe_chr</th>\n",
       "      <th>phe_from</th>\n",
       "      <th>phe_to</th>\n",
       "      <th>phe_strd</th>\n",
       "      <th>n_var_in_cis</th>\n",
       "      <th>dist_phe_var</th>\n",
       "      <th>var_id</th>\n",
       "      <th>var_chr</th>\n",
       "      <th>var_from</th>\n",
       "      <th>var_to</th>\n",
       "      <th>nom_pval</th>\n",
       "      <th>r_squared</th>\n",
       "      <th>slope</th>\n",
       "      <th>best_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [phe_id, phe_chr, phe_from, phe_to, phe_strd, n_var_in_cis, dist_phe_var, var_id, var_chr, var_from, var_to, nom_pval, r_squared, slope, best_hit]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "insQTL CD4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phe_id</th>\n",
       "      <th>phe_chr</th>\n",
       "      <th>phe_from</th>\n",
       "      <th>phe_to</th>\n",
       "      <th>phe_strd</th>\n",
       "      <th>n_var_in_cis</th>\n",
       "      <th>dist_phe_var</th>\n",
       "      <th>var_id</th>\n",
       "      <th>var_chr</th>\n",
       "      <th>var_from</th>\n",
       "      <th>var_to</th>\n",
       "      <th>nom_pval</th>\n",
       "      <th>r_squared</th>\n",
       "      <th>slope</th>\n",
       "      <th>best_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>892144</th>\n",
       "      <td>14397</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110925001</td>\n",
       "      <td>110950000</td>\n",
       "      <td>+</td>\n",
       "      <td>2753</td>\n",
       "      <td>-66474</td>\n",
       "      <td>rs13401811</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858527</td>\n",
       "      <td>110858527</td>\n",
       "      <td>0.009536</td>\n",
       "      <td>0.143105</td>\n",
       "      <td>-0.618911</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>892492</th>\n",
       "      <td>14400</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111000001</td>\n",
       "      <td>111025000</td>\n",
       "      <td>+</td>\n",
       "      <td>2966</td>\n",
       "      <td>-141474</td>\n",
       "      <td>rs13401811</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858527</td>\n",
       "      <td>110858527</td>\n",
       "      <td>0.007402</td>\n",
       "      <td>0.151980</td>\n",
       "      <td>-0.637813</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        phe_id phe_chr   phe_from     phe_to phe_strd  n_var_in_cis  \\\n",
       "892144   14397    chr2  110925001  110950000        +          2753   \n",
       "892492   14400    chr2  111000001  111025000        +          2966   \n",
       "\n",
       "        dist_phe_var      var_id var_chr   var_from     var_to  nom_pval  \\\n",
       "892144        -66474  rs13401811    chr2  110858527  110858527  0.009536   \n",
       "892492       -141474  rs13401811    chr2  110858527  110858527  0.007402   \n",
       "\n",
       "        r_squared     slope  best_hit  \n",
       "892144   0.143105 -0.618911         0  \n",
       "892492   0.151980 -0.637813         0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "insQTL CD8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phe_id</th>\n",
       "      <th>phe_chr</th>\n",
       "      <th>phe_from</th>\n",
       "      <th>phe_to</th>\n",
       "      <th>phe_strd</th>\n",
       "      <th>n_var_in_cis</th>\n",
       "      <th>dist_phe_var</th>\n",
       "      <th>var_id</th>\n",
       "      <th>var_chr</th>\n",
       "      <th>var_from</th>\n",
       "      <th>var_to</th>\n",
       "      <th>nom_pval</th>\n",
       "      <th>r_squared</th>\n",
       "      <th>slope</th>\n",
       "      <th>best_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>924063</th>\n",
       "      <td>14371</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110275001</td>\n",
       "      <td>110300000</td>\n",
       "      <td>+</td>\n",
       "      <td>2430</td>\n",
       "      <td>558527</td>\n",
       "      <td>rs13401811</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858527</td>\n",
       "      <td>110858527</td>\n",
       "      <td>0.003834</td>\n",
       "      <td>0.152552</td>\n",
       "      <td>0.621246</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>925818</th>\n",
       "      <td>14424</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111600001</td>\n",
       "      <td>111625000</td>\n",
       "      <td>+</td>\n",
       "      <td>3628</td>\n",
       "      <td>-741474</td>\n",
       "      <td>rs13401811</td>\n",
       "      <td>chr2</td>\n",
       "      <td>110858527</td>\n",
       "      <td>110858527</td>\n",
       "      <td>0.009671</td>\n",
       "      <td>0.124136</td>\n",
       "      <td>0.560406</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        phe_id phe_chr   phe_from     phe_to phe_strd  n_var_in_cis  \\\n",
       "924063   14371    chr2  110275001  110300000        +          2430   \n",
       "925818   14424    chr2  111600001  111625000        +          3628   \n",
       "\n",
       "        dist_phe_var      var_id var_chr   var_from     var_to  nom_pval  \\\n",
       "924063        558527  rs13401811    chr2  110858527  110858527  0.003834   \n",
       "925818       -741474  rs13401811    chr2  110858527  110858527  0.009671   \n",
       "\n",
       "        r_squared     slope  best_hit  \n",
       "924063   0.152552  0.621246         0  \n",
       "925818   0.124136  0.560406         0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "allelic imbalance\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>combined_p_val_greater</th>\n",
       "      <th>combined_p_val_less</th>\n",
       "      <th>tot_REF</th>\n",
       "      <th>tot_ALT</th>\n",
       "      <th>ratio</th>\n",
       "      <th>n_pat</th>\n",
       "      <th>corrected_p_val_greater</th>\n",
       "      <th>corrected_p_val_less</th>\n",
       "      <th>TF_remap</th>\n",
       "      <th>TF_JASPAR</th>\n",
       "      <th>eQTLgen_symbol</th>\n",
       "      <th>eQTLgen_pval</th>\n",
       "      <th>CD4_lowest_allele_specific</th>\n",
       "      <th>CD8_lowest_allele_specific</th>\n",
       "      <th>ATAC_hic_corr_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25963</th>\n",
       "      <td>chr2</td>\n",
       "      <td>110858527</td>\n",
       "      <td>rs13401811</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.594457e-98</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3724.0</td>\n",
       "      <td>2207.0</td>\n",
       "      <td>0.592642</td>\n",
       "      <td>29.0</td>\n",
       "      <td>1.856879e-96</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{GABPA, JUNB, SPI1, SKIL, FOXM1, ZFP36, BHLHE4...</td>\n",
       "      <td>{ZEB1, SNAI1, NR4A2, GATA4, JUNB, TCF3, Ptf1A,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM        POS          ID REF ALT  combined_p_val_greater  \\\n",
       "25963  chr2  110858527  rs13401811   G   A            1.594457e-98   \n",
       "\n",
       "       combined_p_val_less  tot_REF  tot_ALT     ratio  n_pat  \\\n",
       "25963                  1.0   3724.0   2207.0  0.592642   29.0   \n",
       "\n",
       "       corrected_p_val_greater  corrected_p_val_less  \\\n",
       "25963             1.856879e-96                   1.0   \n",
       "\n",
       "                                                TF_remap  \\\n",
       "25963  {GABPA, JUNB, SPI1, SKIL, FOXM1, ZFP36, BHLHE4...   \n",
       "\n",
       "                                               TF_JASPAR eQTLgen_symbol  \\\n",
       "25963  {ZEB1, SNAI1, NR4A2, GATA4, JUNB, TCF3, Ptf1A,...            NaN   \n",
       "\n",
       "      eQTLgen_pval  CD4_lowest_allele_specific  CD8_lowest_allele_specific  \\\n",
       "25963          NaN                         NaN                         NaN   \n",
       "\n",
       "       ATAC_hic_corr_score  \n",
       "25963                  0.1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loops with allelic imbalance CD4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrA</th>\n",
       "      <th>startA</th>\n",
       "      <th>endA</th>\n",
       "      <th>chrB</th>\n",
       "      <th>startB</th>\n",
       "      <th>endB</th>\n",
       "      <th>loopID</th>\n",
       "      <th>loopScore</th>\n",
       "      <th>rsID</th>\n",
       "      <th>rsCoord</th>\n",
       "      <th>combined_Pval_greater</th>\n",
       "      <th>corrected_Pval_greater</th>\n",
       "      <th>combined_Pval_less</th>\n",
       "      <th>corrected_Pval_less</th>\n",
       "      <th>tot_REF</th>\n",
       "      <th>tot_ALT</th>\n",
       "      <th>n_pat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrA, startA, endA, chrB, startB, endB, loopID, loopScore, rsID, rsCoord, combined_Pval_greater, corrected_Pval_greater, combined_Pval_less, corrected_Pval_less, tot_REF, tot_ALT, n_pat]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loops with allelic imbalance CD8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrA</th>\n",
       "      <th>startA</th>\n",
       "      <th>endA</th>\n",
       "      <th>chrB</th>\n",
       "      <th>startB</th>\n",
       "      <th>endB</th>\n",
       "      <th>loopID</th>\n",
       "      <th>loopScore</th>\n",
       "      <th>rsID</th>\n",
       "      <th>rsCoord</th>\n",
       "      <th>combined_Pval_greater</th>\n",
       "      <th>corrected_Pval_greater</th>\n",
       "      <th>combined_Pval_less</th>\n",
       "      <th>corrected_Pval_less</th>\n",
       "      <th>tot_REF</th>\n",
       "      <th>tot_ALT</th>\n",
       "      <th>n_pat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrA, startA, endA, chrB, startB, endB, loopID, loopScore, rsID, rsCoord, combined_Pval_greater, corrected_Pval_greater, combined_Pval_less, corrected_Pval_less, tot_REF, tot_ALT, n_pat]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "get_all_tables(\"rs13401811\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
