{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# process the individual samples and extract p values\n",
    "\n",
    "we cannot provide raw genotypes for the individual patients, so this script is only for reference\n",
    "\n",
    "the way this script works is that it opens each individually called BCF file. It then extracts the useful information, identifies the heterozygous sites and then calculates a basic p-value for each patient.\n",
    "\n",
    "the next script will merge the p-values per variant and produce the final script"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# note that we cannot provide raw genotypes for the patients, so this script is here only as a reference and it cannot be run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 28 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import hicstraw \n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import glob\n",
    "import os\n",
    "import plotly.express as px\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors\n",
    "from pandarallel import pandarallel\n",
    "import cooler\n",
    "import cooltools\n",
    "import pybedtools as pbed\n",
    "pandarallel.initialize()\n",
    "import scipy\n",
    "from statsmodels.stats import multitest\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import plotly.io as pio\n",
    "import seaborn as sns\n",
    "import gzip\n",
    "import io\n",
    "import pickle\n",
    "from functools import reduce\n",
    "\n",
    "os.makedirs(\"/mnt/iusers01/jw01/mdefscs4/scratch/temp_pybedtools/\", exist_ok = True)\n",
    "pbed.helpers.set_tempdir(\"/mnt/iusers01/jw01/mdefscs4/scratch/temp_pybedtools/\")\n",
    "bed_genome_file = \"/mnt/iusers01/jw01/mdefscs4/hg38.genome\"\n",
    "\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "base_dir = \"/mnt/jw01-aruk-home01/projects/psa_functional_genomics/PsA_cleaned_analysis\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_ATAC = pd.read_csv(f\"{base_dir}/metadata/cleaned_ATAC_metadata.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf(path):\n",
    "    with gzip.open(path, 'rt') as f:\n",
    "        lines = [l for l in f if not l.startswith('##')]\n",
    "    return pd.read_csv(\n",
    "        io.StringIO(''.join(lines)),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "               'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NRHV151_CD4 21886\n",
      "NRHV151_CD8 22192\n",
      "NRHV321_CD4 4027\n",
      "PsA4920_CD4 23011\n",
      "PsA4945_CD8 11650\n",
      "PsA4948_CD8 13267\n",
      "PsA4953_CD8 45741\n",
      "PsA4960_CD8 12760\n",
      "NRHV321_CD8 16892\n",
      "PsA4944_CD4 22369\n",
      "PsA4947_CD8 13527\n",
      "PsA4949_CD4 18293\n",
      "PsA4954_CD8 27711\n",
      "PsA4963_CD8 17832\n",
      "NRHV238_CD4 26499\n",
      "nrhv238_CD8_ 12359\n",
      "PsA4950_CD4 17559\n",
      "PsA4961_CD8 21091\n",
      "PsA4964_CD8 10565\n",
      "PsA5012_CD4 28813\n",
      "NRHV326_CD8 21260\n",
      "PsA4920_CD8 20468\n",
      "PsA4952_CD8 23734\n",
      "PsA4963_CD4 10514\n",
      "PsA5008_CD4 24347\n",
      "PsA5012_CD8 25585\n",
      "PsA4941CD8 20015\n",
      "PsA4944CD8 22952\n",
      "PSA4946CD8 18533\n",
      "PsA4952CD4 10183\n",
      "PsA4954CD4 17933\n",
      "PsA4962CD8 10243\n",
      "PSA4964CD4 12899\n",
      "PsA4965CD8 21027\n",
      "PSA4969CD8 26397\n",
      "PsA4969CD8_SF 20100\n",
      "PSA5008CD8 26988\n",
      "PsA5014CD4 24709\n",
      "PsA5014CD8 27362\n",
      "PsA5019CD8 20563\n",
      "NRHV171CD4 14358\n",
      "NRHV171CD8 14450\n",
      "NRHV324CD4 26733\n",
      "PsA4950CD8 7297\n",
      "PsA4958CD4 21714\n",
      "PsA4958CD8 15293\n",
      "PsA5009CD4 19038\n",
      "PsA5010CD4 28587\n",
      "PsA5010CD8 33084\n",
      "PsA5018CD4 28255\n",
      "PsA5018CD8 25937\n",
      "PsA5023CD8 27508\n",
      "PsA5026CD4 21747\n",
      "PsA5026CD8 25178\n",
      "PsA4956CD4 24624\n",
      "PsA4956CD8 25227\n",
      "PsA4968CD8 24986\n",
      "PsA5007CD8 24347\n",
      "PsA5009CD8 24725\n",
      "PsA5013CD4 21342\n",
      "PsA5013CD8 25395\n",
      "PsA5015CD8 29203\n",
      "PsA5020CD8 16779\n",
      "PsA5021CD8 20926\n",
      "PsA5022CD8 16511\n",
      "PsA5024CD8 19706\n",
      "PsA5036CD4 26945\n",
      "PsA5036CD8 29219\n",
      "NRHV086CD4 22983\n",
      "NRHV086CD8 25440\n",
      "NRHV121CD4 20544\n",
      "NRHV121CD8 21694\n",
      "NRHV322CD4 24436\n",
      "NRHV322CD8 24070\n",
      "NRHV324CD8 24744\n",
      "PSA4947CD4 21088\n",
      "PsA4957CD4 27344\n",
      "PsA4959CD4 19067\n",
      "PsA5006CD8 23116\n",
      "PsA5019CD4 26157\n",
      "PsA5021CD4 18286\n",
      "NRHV073CD4 17488\n",
      "NRHV073CD8 17125\n",
      "NRHV079CD4 20793\n",
      "NRHV079CD8 23283\n",
      "NRHV168CD4 19217\n",
      "NRHV290CD4 16596\n",
      "NRHV295CD4 25337\n",
      "NRHV295CD8 16142\n",
      "NRHV332CD8 15469\n",
      "PSA5040CD8 18634\n",
      "PSA5040CD8SF 19388\n",
      "NRHV168_CD8 20696\n",
      "NRHV290_CD8 27139\n",
      "NRHV326CD4 19344\n",
      "NRHV332_CD4 24965\n",
      "PSA4945_CD4 7283\n",
      "PSA4949_CD8 5119\n",
      "psa4951_CD4 16385\n",
      "psa4951_CD8 15804\n",
      "PSA4959_CD8 20409\n",
      "PSA5017_CD4 26163\n",
      "PSA5017_CD8 22418\n",
      "PSA5025CD8SF 19282\n",
      "PSA5037_CD4 21033\n",
      "PSA5039_CD8 11071\n",
      "NRHV014XCD4 18534\n",
      "NRHV014XCD8 22421\n",
      "NRHV325_CD4 22448\n",
      "NRHV325_CD8 23517\n",
      "NRSF055CD4SF 27850\n",
      "NRSF055CD8SF 24134\n",
      "NRSF056CD8SF 12957\n",
      "PSA4951CD4SF 28884\n",
      "PSA4958CD8SF 9687\n",
      "PSA4967CD4 23037\n",
      "PSA4968CD4 22558\n",
      "PSA5006CD4 25894\n",
      "PSA5007CD4 19582\n",
      "PSA4942CD8SF 38318\n",
      "PSA4955CD8 34190\n",
      "PSA4958CD4SF 21700\n",
      "PSA4962CD4 9503\n",
      "PSA4966CD8 20336\n",
      "PSA5037CD8 30559\n",
      "PSA5040CD4 25836\n"
     ]
    }
   ],
   "source": [
    "def extract_numbers(x):\n",
    "    # split the string by ':'\n",
    "    numbers = x.split(':')\n",
    "    # extract the last two numbers\n",
    "    num1, num2 = numbers[-1].split(',')\n",
    "    return int(num1), int(num2)\n",
    "\n",
    "\n",
    "def get_data_per_sample(ID_sample):\n",
    "    dataframes = []\n",
    "\n",
    "    # loop through the 22 chromosomes\n",
    "    for chrom in range(1, 23):\n",
    "        # open the dataframe for each chromosome\n",
    "        test_file = f\"/mnt/iusers01/jw01/mdefscs4/psa_functional_genomics/PsA_combined_analysis/allele_specific_hic/ATAC_specific/called_bcf/{ID_sample}/{ID_sample}_chr{chrom}_filt_annotated.vcf.gz\"\n",
    "        df = read_vcf(test_file)\n",
    "        # add the dataframe to the list\n",
    "        dataframes.append(df)\n",
    "\n",
    "    # concatenate the dataframes\n",
    "    result = pd.concat(dataframes).reset_index(drop = True)\n",
    "\n",
    "    result = result.drop([\"FILTER\", \"QUAL\", \"INFO\", \"FORMAT\"], axis = 1)\n",
    "    result = result.rename(columns={result.columns[-1]: ID_sample})\n",
    "\n",
    "    result[[ID_sample + '_A', ID_sample + '_B']] = pd.DataFrame(result[ID_sample].apply(extract_numbers).tolist(), columns=[ID_sample + '_A', ID_sample + '_B'])\n",
    "    \n",
    "    # filter rows here\n",
    "    def get_good_row(row):\n",
    "        if (row[ID_sample + '_A'] > 2) and (row[ID_sample + '_B'] > 2): # either of them needs to be above two\n",
    "            tot = row[ID_sample + '_A']+row[ID_sample + '_B']\n",
    "            if (row[ID_sample + '_A']/tot > 0.05) and (row[ID_sample + '_B']/tot > 0.05): # both of them need to be above 5%, this is in case actual counts are very high\n",
    "                return True\n",
    "        return False\n",
    "    result = result[result.apply(get_good_row, axis = 1)]\n",
    "    def retrieve_p_val(row):\n",
    "        row[f\"{ID_sample}_pval_greater\"] = scipy.stats.binom_test(row[f\"{ID_sample}_A\"],row[f\"{ID_sample}_B\"] + row[f\"{ID_sample}_A\"],0.5,\"greater\")\n",
    "        row[f\"{ID_sample}_pval_less\"] = scipy.stats.binom_test(row[f\"{ID_sample}_A\"],row[f\"{ID_sample}_B\"] + row[f\"{ID_sample}_A\"],0.5,\"less\")\n",
    "        return row\n",
    "    result = result.parallel_apply(retrieve_p_val, axis = 1)\n",
    "    return result\n",
    "\n",
    "all_results = {}\n",
    "for sample in metadata_ATAC[\"id\"].to_list():\n",
    "    result = get_data_per_sample(sample)\n",
    "    all_results[sample] = result\n",
    "\n",
    "    print(sample, len(result))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pickle.dump(all_results, open(f\"{base_dir}/ATAC_allelic_imbalance/all_p_val_called.pk\", \"wb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "pickle.dump(all_results, open(\"/mnt/iusers01/jw01/mdefscs4/psa_functional_genomics/PsA_combined_analysis/allele_specific_hic/ATAC_specific/all_p_val_called.pk\", \"wb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
